File: examples/webhook_receiver.js
/**
 * ROJUDGER Webhook Receiver - Node.js + Express
 *
 * Este ejemplo muestra cómo recibir y verificar webhooks de ROJUDGER
 * con validación de firma HMAC-SHA256.
 *
 * Instalación:
 *   npm install express
 *
 * Uso:
 *   WEBHOOK_SECRET="tu-secreto" node webhook_receiver.js
 */

const express = require('express');
const crypto = require('crypto');

const app = express();
const PORT = process.env.PORT || 9000;
const WEBHOOK_SECRET = process.env.WEBHOOK_SECRET || '';

// ⚠️ Importante: usar raw body para verificar HMAC
app.use('/webhooks/rojudger', express.raw({ type: 'application/json' }));
app.use(express.json()); // Para otras rutas

/**
 * Verifica la firma HMAC del webhook
 */
function verifyWebhookSignature(body, signature) {
  if (!WEBHOOK_SECRET) {
    console.warn('⚠️  WEBHOOK_SECRET no configurado. Saltando verificación.');
    return true;
  }

  const hmac = crypto.createHmac('sha256', WEBHOOK_SECRET);
  hmac.update(body);
  const expectedSignature = hmac.digest('hex');

  // Comparación segura contra timing attacks
  return crypto.timingSafeEqual(
    Buffer.from(signature),
    Buffer.from(expectedSignature)
  );
}

/**
 * Handler principal del webhook
 */
app.post('/webhooks/rojudger', (req, res) => {
  const signature = req.headers['x-rojudger-signature'] || '';
  const submissionId = req.headers['x-rojudger-submission-id'];
  const event = req.headers['x-rojudger-event'];
  const body = req.body;

  console.log('\n' + '='.repeat(60));
  console.log(`📨 Webhook recibido: ${new Date().toISOString()}`);
  console.log('='.repeat(60));

  // 1. Verificar firma HMAC
  if (WEBHOOK_SECRET) {
    if (!verifyWebhookSignature(body, signature)) {
      console.error('❌ Firma HMAC inválida!');
      return res.status(401).json({ error: 'Invalid signature' });
    }
    console.log('✅ Firma HMAC verificada');
  }

  // 2. Parsear payload
  let payload;
  try {
    payload = JSON.parse(body);
  } catch (err) {
    console.error('❌ JSON inválido:', err.message);
    return res.status(400).json({ error: 'Invalid JSON' });
  }

  // 3. Validar estructura
  if (!payload.submission || !payload.submission.id) {
    console.error('❌ Payload inválido: falta submission.id');
    return res.status(400).json({ error: 'Invalid payload' });
  }

  const { submission } = payload;

  // 4. Log de información
  console.log(`📋 Submission ID: ${submission.id}`);
  console.log(`🏷️  Event: ${event || payload.event}`);
  console.log(`📊 Status: ${submission.status}`);
  console.log(`🔢 Exit Code: ${submission.exit_code}`);
  console.log(`⏱️  Time: ${submission.time}s`);
  console.log(`💾 Memory: ${submission.memory} KB`);

  if (submission.stdout) {
    console.log(`📤 Stdout:\n${submission.stdout.substring(0, 200)}`);
  }

  if (submission.stderr) {
    console.log(`⚠️  Stderr:\n${submission.stderr.substring(0, 200)}`);
  }

  if (submission.compile_output) {
    console.log(`🔧 Compile Output:\n${submission.compile_output.substring(0, 200)}`);
  }

  if (submission.message) {
    console.log(`💬 Message: ${submission.message}`);
  }

  // 5. Procesar según status
  switch (submission.status) {
    case 'completed':
      handleCompletedSubmission(submission);
      break;
    case 'error':
      handleErrorSubmission(submission);
      break;
    case 'timeout':
      handleTimeoutSubmission(submission);
      break;
    default:
      console.warn(`⚠️  Status desconocido: ${submission.status}`);
  }

  // 6. Responder rápidamente
  res.status(200).json({
    status: 'received',
    submission_id: submission.id,
    timestamp: new Date().toISOString()
  });

  console.log('✅ Webhook procesado correctamente\n');
});

/**
 * Procesar submission completada exitosamente
 */
function handleCompletedSubmission(submission) {
  console.log('🎉 Submission completada exitosamente');

  // Aquí puedes:
  // - Actualizar base de datos local
  // - Enviar notificación al usuario
  // - Calcular estadísticas
  // - Actualizar leaderboard
  // - etc.

  // Ejemplo: guardar en DB (pseudo-código)
  // db.submissions.update({
  //   id: submission.id,
  //   status: 'completed',
  //   result: submission.stdout,
  //   time: submission.time
  // });
}

/**
 * Procesar submission con error
 */
function handleErrorSubmission(submission) {
  console.log('❌ Submission falló');

  // Aquí puedes:
  // - Notificar al usuario del error
  // - Registrar para debugging
  // - Ofrecer retry automático
  // - etc.

  if (submission.compile_output) {
    console.log('Falló en compilación');
  } else if (submission.exit_code !== 0) {
    console.log('Error en runtime');
  }
}

/**
 * Procesar submission con timeout
 */
function handleTimeoutSubmission(submission) {
  console.log('⏱️  Submission excedió tiempo límite');

  // Aquí puedes:
  // - Notificar al usuario
  // - Sugerir optimización
  // - etc.
}

/**
 * Health check endpoint
 */
app.get('/health', (req, res) => {
  res.json({
    status: 'healthy',
    timestamp: new Date().toISOString(),
    webhook_secret_configured: !!WEBHOOK_SECRET
  });
});

/**
 * Endpoint de prueba manual
 */
app.post('/test-webhook', express.json(), (req, res) => {
  console.log('🧪 Test webhook recibido:', req.body);
  res.json({ status: 'test received' });
});

/**
 * Iniciar servidor
 */
app.listen(PORT, () => {
  console.log('╔════════════════════════════════════════════════════╗');
  console.log('║     ROJUDGER Webhook Receiver (Node.js)            ║');
  console.log('╚════════════════════════════════════════════════════╝');
  console.log(`🚀 Servidor escuchando en http://0.0.0.0:${PORT}`);
  console.log(`🔒 HMAC Secret: ${WEBHOOK_SECRET ? '✅ Configurado' : '❌ No configurado'}`);
  console.log('');
  console.log('📡 Endpoints:');
  console.log(`   POST http://localhost:${PORT}/webhooks/rojudger`);
  console.log(`   GET  http://localhost:${PORT}/health`);
  console.log('');
  console.log('💡 Para testear:');
  console.log('   curl -X POST http://localhost:8080/api/v1/submissions \\');
  console.log('     -d \'{"language_id": 71, "source_code": "print(\\"test\\")", \\');
  console.log(`          "webhook_url": "http://localhost:${PORT}/webhooks/rojudger"}\'`);
  console.log('');
  console.log('⏳ Esperando webhooks...\n');
});

// Manejo de errores global
process.on('uncaughtException', (err) => {
  console.error('❌ Uncaught Exception:', err);
});

process.on('unhandledRejection', (reason, promise) => {
  console.error('❌ Unhandled Rejection at:', promise, 'reason:', reason);
});

// Graceful shutdown
process.on('SIGINT', () => {
  console.log('\n👋 Deteniendo servidor...');
  process.exit(0);
});

module.exports = app; // Para testing

File: examples/webhook_receiver.py
#!/usr/bin/env python3
"""
ROJUDGER Webhook Receiver - Python + Flask

Este ejemplo muestra cómo recibir y verificar webhooks de ROJUDGER
con validación de firma HMAC-SHA256.

Instalación:
    pip install flask

Uso:
    WEBHOOK_SECRET="tu-secreto" python webhook_receiver.py
"""

import os
import hmac
import hashlib
import json
from datetime import datetime
from flask import Flask, request, jsonify

app = Flask(__name__)
WEBHOOK_SECRET = os.getenv('WEBHOOK_SECRET', '').encode()
PORT = int(os.getenv('PORT', 9000))


def verify_webhook_signature(body: bytes, signature: str) -> bool:
    """
    Verifica la firma HMAC del webhook

    Args:
        body: Raw body del request (bytes)
        signature: Firma recibida en header

    Returns:
        True si la firma es válida, False en caso contrario
    """
    if not WEBHOOK_SECRET:
        print('⚠️  WEBHOOK_SECRET no configurado. Saltando verificación.')
        return True

    expected = hmac.new(WEBHOOK_SECRET, body, hashlib.sha256).hexdigest()

    # Comparación segura contra timing attacks
    return hmac.compare_digest(signature, expected)


def handle_completed_submission(submission: dict):
    """Procesar submission completada exitosamente"""
    print('🎉 Submission completada exitosamente')

    # Aquí puedes:
    # - Actualizar base de datos local
    # - Enviar notificación al usuario
    # - Calcular estadísticas
    # - Actualizar leaderboard
    # - etc.

    # Ejemplo: guardar en DB (pseudo-código)
    # db.submissions.update(
    #     id=submission['id'],
    #     status='completed',
    #     result=submission['stdout'],
    #     time=submission['time']
    # )


def handle_error_submission(submission: dict):
    """Procesar submission con error"""
    print('❌ Submission falló')

    # Aquí puedes:
    # - Notificar al usuario del error
    # - Registrar para debugging
    # - Ofrecer retry automático
    # - etc.

    if submission.get('compile_output'):
        print('Falló en compilación')
    elif submission.get('exit_code', 0) != 0:
        print('Error en runtime')


def handle_timeout_submission(submission: dict):
    """Procesar submission con timeout"""
    print('⏱️  Submission excedió tiempo límite')

    # Aquí puedes:
    # - Notificar al usuario
    # - Sugerir optimización
    # - etc.


@app.route('/webhooks/rojudger', methods=['POST'])
def webhook_handler():
    """Handler principal del webhook"""

    signature = request.headers.get('X-Rojudger-Signature', '')
    submission_id = request.headers.get('X-Rojudger-Submission-Id')
    event = request.headers.get('X-Rojudger-Event')
    body = request.get_data()

    print('\n' + '=' * 60)
    print(f'📨 Webhook recibido: {datetime.now().isoformat()}')
    print('=' * 60)

    # 1. Verificar firma HMAC
    if WEBHOOK_SECRET:
        if not verify_webhook_signature(body, signature):
            print('❌ Firma HMAC inválida!')
            return jsonify({'error': 'Invalid signature'}), 401
        print('✅ Firma HMAC verificada')

    # 2. Parsear payload
    try:
        payload = json.loads(body)
    except json.JSONDecodeError as e:
        print(f'❌ JSON inválido: {e}')
        return jsonify({'error': 'Invalid JSON'}), 400

    # 3. Validar estructura
    if 'submission' not in payload or 'id' not in payload.get('submission', {}):
        print('❌ Payload inválido: falta submission.id')
        return jsonify({'error': 'Invalid payload'}), 400

    submission = payload['submission']

    # 4. Log de información
    print(f"📋 Submission ID: {submission['id']}")
    print(f"🏷️  Event: {event or payload.get('event')}")
    print(f"📊 Status: {submission['status']}")
    print(f"🔢 Exit Code: {submission.get('exit_code')}")
    print(f"⏱️  Time: {submission.get('time')}s")
    print(f"💾 Memory: {submission.get('memory')} KB")

    if submission.get('stdout'):
        print(f"📤 Stdout:\n{submission['stdout'][:200]}")

    if submission.get('stderr'):
        print(f"⚠️  Stderr:\n{submission['stderr'][:200]}")

    if submission.get('compile_output'):
        print(f"🔧 Compile Output:\n{submission['compile_output'][:200]}")

    if submission.get('message'):
        print(f"💬 Message: {submission['message']}")

    # 5. Procesar según status
    status = submission.get('status')
    if status == 'completed':
        handle_completed_submission(submission)
    elif status == 'error':
        handle_error_submission(submission)
    elif status == 'timeout':
        handle_timeout_submission(submission)
    else:
        print(f"⚠️  Status desconocido: {status}")

    # 6. Responder rápidamente
    print('✅ Webhook procesado correctamente\n')

    return jsonify({
        'status': 'received',
        'submission_id': submission['id'],
        'timestamp': datetime.now().isoformat()
    }), 200


@app.route('/health', methods=['GET'])
def health_check():
    """Health check endpoint"""
    return jsonify({
        'status': 'healthy',
        'timestamp': datetime.now().isoformat(),
        'webhook_secret_configured': bool(WEBHOOK_SECRET)
    })


@app.route('/test-webhook', methods=['POST'])
def test_webhook():
    """Endpoint de prueba manual"""
    print('🧪 Test webhook recibido:', request.get_json())
    return jsonify({'status': 'test received'})


@app.errorhandler(Exception)
def handle_error(error):
    """Manejo global de errores"""
    print(f'❌ Error: {error}')
    return jsonify({'error': str(error)}), 500


def print_banner():
    """Imprime banner de inicio"""
    print('╔════════════════════════════════════════════════════╗')
    print('║     ROJUDGER Webhook Receiver (Python)             ║')
    print('╚════════════════════════════════════════════════════╝')
    print(f'🚀 Servidor escuchando en http://0.0.0.0:{PORT}')
    print(f"🔒 HMAC Secret: {'✅ Configurado' if WEBHOOK_SECRET else '❌ No configurado'}")
    print('')
    print('📡 Endpoints:')
    print(f'   POST http://localhost:{PORT}/webhooks/rojudger')
    print(f'   GET  http://localhost:{PORT}/health')
    print('')
    print('💡 Para testear:')
    print('   curl -X POST http://localhost:8080/api/v1/submissions \\')
    print('     -d \'{"language_id": 71, "source_code": "print(\\"test\\")", \\')
    print(f'          "webhook_url": "http://localhost:{PORT}/webhooks/rojudger"}\'')
    print('')
    print('⏳ Esperando webhooks...\n')


if __name__ == '__main__':
    print_banner()

    # Ejecutar servidor
    app.run(
        host='0.0.0.0',
        port=PORT,
        debug=False  # Cambiar a True para desarrollo
    )
